reactor-extra/src/main/java/reactor/retry/RetryBuilder.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (c) 2017 Pivotal Software Inc, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.retry;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import org.reactivestreams.Publisher;
+
+import reactor.core.publisher.EmitterProcessor;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Scheduler;
+
+/**
+ * Retry utility to control retry or repeat of {@link Flux} or @link {@link Mono} using
+ * <ul>
+ *   <li>{@link Flux#retryWhen(Function)} </li>
+ *   <li>{@link Flux#repeatWhen(Function)} </li>
+ *   <li>{@link Mono#retryWhen(Function)} </li>
+ *   <li>{@link Mono#repeatWhen(Function)} </li>
+ *   <li>{@link Mono#repeatWhenEmpty(Function)} </li>
+ *   <li>{@link Mono#repeatWhenEmpty(int, Function)} </li>
+ * </ul>
+ *
+ * This class enables commonly used retry/repeat patterns:
+ * <ul>
+ *   <li>Configured number of retries without backoff</li>
+ *   <li>Configured number of retries with fixed backoff interval</li>
+ *   <li>Exponential backoff with timeout</li>
+ *   <li>Retriable exception configuration that can be used with/without backoff</li>
+ *   <li>Non-retriable exception configuration that can be used with/without backoff</li>
+ *   <li>doOnRetry to perform any rollback operation before performing retry</li>
+ * </ul>
+ *
+ * This class does not support fully flexible functions which can be done just as easily on {@link Flux#retryWhen(Function)} etc.
+ * <ul>
+ *   <li>Generic predicate for retry decision</li>
+ *   <li>Flexible delay based on context</li>
+ * </ul>
+ *
+ */
+public class RetryBuilder {
+
+	static final Duration NO_RETRY = Duration.ofSeconds(-1);
+
+	final Predicate<Throwable> retryPredicate;
+	Integer maxAttempts;
+	Duration timeout;
+	Instant timeoutInstant;
+	Function<Long, Duration> backoffCalculator;
+	Scheduler backoffScheduler;
+	Consumer<RetryContext> onRetry;
+
+
+	RetryBuilder(Predicate<Throwable> retryPredicate) {
+		this.retryPredicate = retryPredicate;
+		this.maxAttempts = null;
+		this.backoffCalculator = iteration -> Duration.ZERO;
```

Comment:
oops missed that: probably a bad default since one can call `maxAttempts()` without calling `noBackoff()` (or any explicit backoff strategy method). Should probably default to the same behavior as `noBackoff()`.